### PR TITLE
Roll back changes that allowed setState to be called inside willUpdate

### LINKED ISSUE #153: Roll back `willUpdate` part of #139
#139 loosened the restrictions of `setState` in both `willUpdate` and `init`. Doing the latter was a good change as it reduced confusion and disparity between assigning to `state` directly but only ever in `init` and calling `setState` everywhere else.

Making `willUpdate` allow `setState` was a grave mistake I think, however. I ran into this when working on the `new-reconciler` branch and reimplementing the `setState` variants. Should the new (not yet committed) state be used for state merges? Should a new update happen since `shouldUpdate` is getting skipped?

The case that came up internally that prompted me to consider this change would've been better solved by `getDerivedStateFromProps`. It should be adjusted to match current behavior in React 16.3, which would've addressed the same issue in a more elegant way.

### PR DESCRIPTION
**Pull request by ZoteTheMighty**
Closes #153.

Rolls back the changes to `Component` that allowed `setState` to be called during `willUpdate` . This helps avoid some edge cases in state updates where correct behavior was difficult to determine.

Checklist before submitting:
* [x] Added entry to `CHANGELOG.md`
* [x] Added/updated relevant tests
* [x] Added/updated documentation

### CODE REVIEW

File: lib/Component.spec.lua [hunk 1]
```diff
@@ -456,17 +454,12 @@ return function()
...
-			forceUpdate()
-
-			expect(renderCount).to.equal(2)
+			local element = createElement(TestComponent)
 
-			Reconciler.unmount(handle)
+			expect(function()
+				Reconciler.mount(element)
+				forceUpdate()
+			end).to.throw()
```
Comment (LPGhatguy) [inline lib/Component.spec.lua]: This test will also pass if we rename `_forceUpdate`, an internal API. Can we refactor this test to use `Reconciler.reconcile` or whatnot instead?

Comment (ZoteTheMighty) [reply lib/Component.spec.lua]: Good point, I'll make that change.
